bauble/src/parse/value.rs

```diff
@@ -0,0 +1,375 @@
+use core::fmt;
+
+use crate::spanned::{Span, SpanExt, Spanned};
+use indexmap::IndexMap;
+use rust_decimal::Decimal;
+
+// Could maybe use `&str`?
+pub type Ident = Spanned<String>;
+
+pub type Map = Vec<(Object, Object)>;
+
+pub type Fields = IndexMap<Ident, Object>;
+
+pub type Sequence = Vec<Object>;
+
+pub type Use = Spanned<PathTreeNode>;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum PathEnd {
+    /// path::*::ident
+    WithIdent(Ident),
+    /// path::ident
+    Ident(Ident),
+}
+
+impl fmt::Display for PathEnd {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            PathEnd::WithIdent(ident) => write!(f, "*::{ident}"),
+            PathEnd::Ident(ident) => write!(f, "{ident}"),
+        }
+    }
+}
+
+#[derive(PartialEq, Clone)]
+pub struct Path {
+    pub leading: Spanned<Vec<Ident>>,
+    pub last: Spanned<PathEnd>,
+}
+
+impl Path {
+    pub fn is_just(&self, s: &str) -> bool {
+        self.leading.is_empty()
+            && match &self.last.value {
+                PathEnd::WithIdent(_) => false,
+                PathEnd::Ident(ident) => ident.as_str() == s,
+            }
+    }
+
+    pub fn as_ident(&self) -> Option<&str> {
+        if let (true, PathEnd::Ident(ident)) = (self.leading.is_empty(), &self.last.value) {
+            Some(ident.as_str())
+        } else {
+            None
+        }
+    }
+}
+
+impl fmt::Display for Path {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        for leading in self.leading.iter() {
+            write!(f, "{leading}::")?;
+        }
+        write!(f, "{}", self.last)
+    }
+}
+
+impl fmt::Debug for Path {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        <Self as fmt::Display>::fmt(self, f)
+    }
+}
+
+impl Path {
+    pub fn join(&mut self, part: Ident) -> bool {
+        match &mut self.last.value {
+            PathEnd::WithIdent(_) => false,
+            PathEnd::Ident(ident) => {
+                self.leading.push(std::mem::replace(ident, part));
+                true
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Value {
+    Unit,
+    Num(Decimal),
+    Bool(bool),
+    Str(String),
+    Ref(Spanned<Path>),
+
+    Path(Spanned<Path>),
+
+    Map(Map),
+    Struct {
+        name: Option<Spanned<Path>>,
+        fields: Fields,
+    },
+    Tuple {
+        name: Option<Spanned<Path>>,
+        fields: Sequence,
+    },
+    Array(Sequence),
+    Or(Vec<Spanned<Path>>),
+    Raw(String),
+
+    Error,
+}
+
+// Is 3 spaces intended?
```

Comment:
Is it? :thinking:

Reply:
It's basically a pretty print, so depends on how many spaces we want per indent. Maybe it should be configurable, but that can be changed at a later point.

Reply:
I feel like it makes the most sense for it to be 4 spaces, since that's what Rust does.`

Reply:
Changed it to 4